### PR TITLE
Do not return broken processes to the query process pool

### LINKED ISSUE #2962: ["error","list_error","not a row 'reset'"]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description

Calling a list function, throws errors. I initially started noticing these error logs. The unusual thing is after refreshing the endpoints the results come back. I talked to rnewson from couchdb and he thinks the issue sounds like a list function runs but leaves the view server (a couchjs process) in "list" mode, then the next reset (for a map/reduce or something) blows up, as we don't expect it them.

```
[debug] 2020-06-23T03:12:00.194889Z couchdb@127.0.0.1 <0.23278.930> -------- OS Process #Port<0.1239277> Input  :: ["reset",{"reduce_limit":true,"timeout":5000}]
[debug] 2020-06-23T03:12:00.196258Z couchdb@127.0.0.1 <0.23278.930> -------- OS Process #Port<0.1239277> Output :: ["error","list_error","not a row 'reset'"]
[error] 2020-06-23T03:12:00.196567Z couchdb@127.0.0.1 <0.10822.931> 6baa7e28d2 OS Process Error <0.23278.930> :: {<<"list_error">>,<<"not a row 'reset'">>}
[info] 2020-06-23T03:12:00.196771Z couchdb@127.0.0.1 <0.27104.924> -------- couch_proc_manager <0.23278.930> died normal

[debug] 2020-06-23T03:56:01.381814Z couchdb@127.0.0.1 <0.10570.941> -------- OS Process #Port<0.1263125> Output :: ["error","list_error","not a row 'ddoc'"]
[error] 2020-06-23T03:56:01.382309Z couchdb@127.0.0.1 <0.20004.938> -------- OS Process Error <0.10570.941> :: {<<"list_error">>,<<"not a row 'ddoc'">>}
[error] 2020-06-23T03:56:01.385312Z couchdb@127.0.0.1 <0.20004.938> -------- gen_server couch_proc_manager terminated with reason: bad return value {<<"list_error">>,<<"not a row 'ddoc'">>}
--
[debug] 2020-06-23T03:56:16.519341Z couchdb@127.0.0.1 <0.1703.942> -------- OS Process #Port<0.1255745> Input  :: ["reset",{"reduce_limit":true,"timeout":25000}]
[debug] 2020-06-23T03:56:16.519741Z couchdb@127.0.0.1 <0.1703.942> -------- OS Process #Port<0.1255745> Output :: ["error","list_error","not a row 'reset'"]
[error] 2020-06-23T03:56:16.519924Z couchdb@127.0.0.1 <0.11889.942> 598892c5fa OS Process Error <0.1703.942> :: {<<"list_error">>,<<"not a row 'reset'">>}
```

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT give us access or passwords to your actual CouchDB! )

## Steps to Reproduce

[NOTE]: # ( Include commands to reproduce, if possible. curl is preferred. )

## Expected Behaviour

[NOTE]: # ( Tell us what you expected to happen. )

## Your Environment

[TIP]:  # ( Include as many relevant details about your environment as possible. )
[TIP]:  # ( You can paste the output of curl http://YOUR-COUCHDB:5984/ here. )

* CouchDB version used: 
2.3.0
* Operating system and version:
Debian GNU/Linux 9 (stretch) running on Google GCE.

## Additional Context

Erlang: 
20

[TIP]:  # ( Add any other context about the problem here. )

### PR DESCRIPTION
**Pull request by nickva**
Previously, if an error was thrown in a `with_ddoc_proc/2` callback, the process was still returned to the process pool in the `after` clause. However, in some cases, for example when processing a `_list` response, the process might
end up stuck in a bad state state, such that it could not be re-used anymore. In such a case, a subsequent user of that `couch_js` process would end up throwing an error and crashing.

Fixes #2962

## Checklist

- [x] Code is written and works correctly
- [x] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] A PR for documentation changes has been made in https://github.com/apache/couchdb-documentation

### CODE REVIEW

File: src/couch/test/eunit/couchdb_os_proc_pool.erl [hunk 1]
```diff
@@ -205,6 +208,46 @@ should_reduce_pool_on_idle_os_procs() ->
...
+
+        ?assertEqual(0, couch_proc_manager:get_proc_count()),
+        ok = couch_query_servers:with_ddoc_proc(DDoc, fun(_) -> ok end),
+        ?assertEqual(0, meck:num_calls(couch_os_process, stop, 1)),
+        ?assertEqual(1, couch_proc_manager:get_proc_count()),
+
+        ?assertError(bad, couch_query_servers:with_ddoc_proc(DDoc, fun(_) ->
+            error(bad)
+        end)),
+        ?assertEqual(1, meck:num_calls(couch_os_process, stop, 1)),
```
Comment (davisp) [inline src/couch/test/eunit/couchdb_os_proc_pool.erl:243]: Seems like we should add an assertion that there are zero os processes here?

File: src/couch/test/eunit/couchdb_os_proc_pool.erl [hunk 2]
```diff
@@ -205,6 +208,46 @@ should_reduce_pool_on_idle_os_procs() ->
...
+
+        ok = couch_query_servers:with_ddoc_proc(DDoc, fun(_) -> ok end),
```
Comment (davisp) [inline src/couch/test/eunit/couchdb_os_proc_pool.erl:245]: I'm pretty sure the original bug required a second design doc to be sent to the process in this part to trigger the bug, should we do that to be sure? Should be able to just tweak the id or revs of DDoc.

**Review by davisp — CHANGES_REQUESTED** [review]
Couple tweaks to the test but otherwise looks good!

File: src/couch/test/eunit/couchdb_os_proc_pool.erl (hunk 1 above)
Comment (nickva) [reply src/couch/test/eunit/couchdb_os_proc_pool.erl:243]: Good idea

File: src/couch/test/eunit/couchdb_os_proc_pool.erl (hunk 2 above)
Comment (nickva) [reply src/couch/test/eunit/couchdb_os_proc_pool.erl:245]: We could just for completeness, though we already asserted that stop was called to kill the one

**Review by davisp — APPROVED** [review]
+1

**Pull request merged**